controlplane/kubeadm/internal/controllers/controller.go

```diff
@@ -791,6 +810,88 @@ func (r *KubeadmControlPlaneReconciler) reconcileEtcdMembers(ctx context.Context
 	return nil
 }
 
+func (r *KubeadmControlPlaneReconciler) reconcilePreTerminateHook(ctx context.Context, controlPlane *internal.ControlPlane) (ctrl.Result, error) {
+	if !controlPlane.HasDeletingMachine() {
+		return ctrl.Result{}, nil
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+
+	// Pick the Machine with the oldest deletionTimestamp to keep this function deterministic / reentrant
+	// so we only remove the pre-terminate hook from one Machine at a time.
+	deletingMachine := controlPlane.DeletingMachines().OldestDeletionTimestamp()
```

Comment:
it can be null if the list is 0, should this function check that and return early with an error?

Reply:
The list can't have length 0 because otherwise !controlPlane.HasDeletingMachine woud return true

Reply:
At this point we know it isn't of size 0, it was checked on L814

Reply:
We considered this, but we already check for nil in the function inside. When adding it here, we would have duplicated the same to here.